adr/telemetry.md

```diff
@@ -0,0 +1,75 @@
+# Title [ADR-1]
+
+Collecting structured, event based diagnostics (Telemetry)
+
+## Status
+
+Proposed
+
+## Context
+
+The project uses [tracing](https://docs.rs/tracing/latest/tracing/) to
+collect this information. The tracing crate provides the helpful
+"instrument" macro for creating/entering tracing spans on function
+invoke.
+
+Currently, our use of this macro is undifferentiated and we are seeing
+issues where our telemetry data is affecting performance of the router
+because of volumes and format (JSON) of collected data.
+
+## Decision
+
+This ADR proposes that we change our approach to using #[instrument]
+so that we only trace the exact information which we need. This
+will require a review of existing usage.
+
+Recommendations:
+ - Always specify `skip_all` when using the #[instrument] macro
+
+e.g.:
+```
+#[instrument(skip_all, ...)]
+```
+
+This will have the effect of making the macro "opt-in" rather than the
+default behaviour of "opt-out" and will make mistakes easier to spot
+and prevent in code review. This prevents the default serialization of
+all parameters in a span.
+
+Required fields can be specified using the "fields" attribute. See [instrument documentation](https://docs.rs/tracing-attributes/0.1.18/tracing_attributes/attr.instrument.html) for more details.
+
+ - Ensure that each instrumented function is named to promote understanding
+   and consistency
+
+ - Ensure that instrumentation is at the "info" level. Other levels, such
```

Comment:
using instrument at debug or trace level should be fine, it's actually a good use of that attribute, because when debugging we may want to know which function is called. While info level spans are collected in production telemetry, where that level of detail is less interesting, and we actually want logic spans (like "query planning phase")

Reply:
I think that's really the point I'm trying to make here, but perhaps not very clearly. I'm trying to indicate that most instrumentation should be considered as "production first" (i.e. info level). I want developers to think about the level and have a bias for thinking about how to report production useful telemetry.